modules/ibc-hooks/README.md

```diff
@@ -0,0 +1,351 @@
+
+### Implementation
+
+#### Callback information in memo
+
+For the callback to be processed, the transfer packet's `memo` should contain the following in its JSON:
+
+```json
+{"ibc_callback": "osmo1contractAddr"}
+```
+
+The IBC hooks will keep the mapping from the packet's channel and sequence to the contract in storage. When an `Ack` is
+received, it will notify the specified contract via a sudo message.
+
+#### Interface for receiving the Acks and Timeouts
+
+The contract that awaits the callback should implement the following interface for a sudo message:
+
+```rust
+#[cw_serde]
+pub enum IBCLifecycleComplete {
+    #[serde(rename = "ibc_ack")]
+    IBCAck {
+        /// The source channel (osmosis side) of the IBC packet
+        channel: String,
+        /// The sequence number that the packet was sent with
+        sequence: u64,
+        /// String encoded version of the `Ack` as seen by OnAcknowledgementPacket(..)
+        ack: String,
+        /// Weather an `Ack` is a success of failure according to the transfer spec
+        success: bool,
+    },
+    #[serde(rename = "ibc_timeout")]
+    IBCTimeout {
+        /// The source channel (osmosis side) of the IBC packet
+        channel: String,
+        /// The sequence number that the packet was sent with
+        sequence: u64,
+    },
+}
+
+/// Message type for `sudo` entry_point
+#[cw_serde]
+pub enum SudoMsg {
+    #[serde(rename = "ibc_lifecycle_complete")]
+    IBCLifecycleComplete(IBCLifecycleComplete),
+}
+```
+
+## Installation
+
+Follow these steps to install the IBC hooks module. The following lines are all added to `app.go`
+
+1. Import the following packages. 
+
+```go
+// import (
+    ...
+    ibchooks "github.com/cosmos/ibc-apps/modules/ibc-hooks/v7"
+	ibchookskeeper "github.com/cosmos/ibc-apps/modules/ibc-hooks/v7/keeper"
+	ibchookstypes "github.com/cosmos/ibc-apps/modules/ibc-hooks/v7/types"
+    ...
+// )
+```
+
+2. Add the module. 
+
+```go
+// var (
+	//DefaultNodeHome string
+	//ModuleBasics = module.NewBasicManager(
+        ...
+        ibchooks.AppModuleBasic{},
+        ...
+    // ) 
+```
+
+3. Add the IBC hooks keeper. 
+
+```go
+
+// type App struct {
+    ...
+    IBCHooksKeeper   ibchookskeeper.Keeper
+    ...
+// }
+
+```
+4. Initiate the keepers and helpers.
+
+```go
+...
+
+	// 'ibc-hooks' module - depends on
+	// 1. 'auth'
+	// 2. 'bank'
+	// 3. 'distr'
+	app.keys[ibchookstypes.StoreKey] = storetypes.NewKVStoreKey(ibchookstypes.StoreKey)
+	app.IBCHooksKeeper = ibchookskeeper.NewKeeper(
+		app.keys[ibchookstypes.StoreKey],
+	)
+	app.Ics20WasmHooks = ibchooks.NewWasmHooks(&app.IBCHooksKeeper, nil, AccountAddressPrefix) // The contract keeper needs to be set later
```

Comment:
I think a step where we set the contract keeper (implied by the comment in this line) is missing.

it would look something like 

```
        // initialize the wasm keeper with
        // wasmKeeper := wasm.NewKeeper( ...
        //
	app.WasmKeeper = &wasmKeeper

	// Pass the contract keeper to all the structs (generally ICS4Wrappers for ibc middlewares) that need it
	app.ContractKeeper = wasmkeeper.NewDefaultPermissionKeeper(appKeepers.WasmKeeper)
	app.Ics20WasmHooks.ContractKeeper = appKeepers.ContractKeeper
```

Also a potential cleanup here is making the contractKeeper private and adding a SetContractKeeper() method